lib/data/datasource/remote/api_service.dart

```diff
@@ -0,0 +1,137 @@
+import 'dart:convert';
+import 'dart:io';
+
+import 'package:http/http.dart' as http;
+
+import '../../../core/util/mapper/mapper.dart';
+import '../../../domain/entity/genre.dart';
+import '../../../domain/entity/movie.dart';
+import '../../model/data_state.dart';
+import '../../model/genre_model.dart';
+
+import '../../model/now_playing_movies_model.dart';
+import '../../model/popular_movies_model.dart';
+import '../../model/top_rated_movie_model.dart';
+import '../../model/upcoming_movies_model.dart';
+
+class ApiService {
```

Comment:
You are using the same code for 4 different calls (getPopularMovies, getTopRatedMovies, getNowPlayingMovies, getUpcomingMovies), the only thing that it changes is the url that you are calling, so maybe you can use only one method add send the url as a parameter like:
`Future<DataState<List<Movie>>> getMovies({required String url})`

Reply:
done! thanks for this!